File: code/python/05-context-assembly/session_manager.py
"""Session Manager — multi-session conversation management with expiry,
branching, and reset.

Handles the full lifecycle of a conversation session: creation, context
accumulation, branch (inheriting context from a parent), explicit reset, and
expiry after a configurable period of inactivity.

Key classes:

- :class:`Session` — a single user conversation (state + messages + summarizer).
- :class:`SessionManager` — manages a collection of sessions, with TTL and
  persistence helpers.

See: docs/04-context-engineering/04-multi-turn-context-management.md
"""

from __future__ import annotations

import json
import time
import uuid
from dataclasses import dataclass, field
from typing import Optional

from context_budget import count_tokens
from progressive_summarizer import ProgressiveSummarizer
from state_manager import ConversationState, StateManager


# ---------------------------------------------------------------------------
# Session
# ---------------------------------------------------------------------------


class Session:
    """A single conversation session for one user.

    Owns a :class:`~state_manager.ConversationState`,
    a raw ``messages`` list for the LLM, and a
    :class:`~progressive_summarizer.ProgressiveSummarizer` for context
    compression.

    Args:
        user_id: Application-level user identifier (e.g. UUID, email hash).
    """

    def __init__(self, user_id: str) -> None:
        self.session_id: str = str(uuid.uuid4())
        self.user_id: str = user_id
        self.created_at: float = time.time()
        self.last_activity: float = time.time()
        self.state_manager: StateManager = StateManager()
        self.messages: list[dict] = []
        self.summarizer: ProgressiveSummarizer = ProgressiveSummarizer()
        self.inherited_context: Optional[str] = None
        self.is_active: bool = True

    # Convenience property so callers can access state directly
    @property
    def state(self) -> ConversationState:
        return self.state_manager.state

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def is_expired(self, ttl_minutes: int) -> bool:
        """Return True when the session has been idle longer than *ttl_minutes*."""
        return (time.time() - self.last_activity) > (ttl_minutes * 60)

    def touch(self) -> None:
        """Update last-activity timestamp."""
        self.last_activity = time.time()

    # ------------------------------------------------------------------
    # Message management
    # ------------------------------------------------------------------

    def add_user_message(self, message: str) -> None:
        """Append a user message and update conversation state.

        Args:
            message: Raw text of the user's message.
        """
        self.touch()
        self.messages.append({"role": "user", "content": message})
        self.state_manager.process_user_turn(message)

    def add_agent_message(
        self, message: str, tool_calls: list | None = None
    ) -> None:
        """Append an agent message and update conversation state.

        Args:
            message:    Raw text of the agent's response.
            tool_calls: Optional list of tool-call objects for subtask tracking.
        """
        self.touch()
        msg: dict = {"role": "assistant", "content": message}
        if tool_calls:
            msg["tool_calls"] = tool_calls
        self.messages.append(msg)
        self.state_manager.process_agent_turn(message, tool_calls=tool_calls)

        # Feed the turn pair to the progressive summarizer
        # Find the most recent user message
        user_content = ""
        for m in reversed(self.messages[:-1]):
            if m["role"] == "user":
                user_content = m.get("content", "")
                break
        self.summarizer.add_turn(user_content, message)

    def build_messages_for_llm(
        self,
        system_prompt: str = "",
        max_tokens: int = 100_000,
    ) -> list[dict]:
        """Build the messages array for the next LLM call.

        Performs token-aware truncation of the raw message history.
        Injects conversation state and progressive summaries into the
        system prompt.

        Args:
            system_prompt: Base system prompt text.
            max_tokens:    Hard cap on total tokens in the returned list.

        Returns:
            A ``list[dict]`` of ``{"role": ..., "content": ...}`` messages
            ready to pass directly to the chat completions API.
        """
        # Build an augmented system prompt
        augmented = self.state_manager.build_system_prompt_with_state(system_prompt)
        summary_context = self.summarizer.get_context()
        if summary_context:
            augmented = (
                f"{augmented}\n\n"
                f"## Conversation History Summary\n"
                f"{summary_context}"
            )
        if self.inherited_context:
            augmented = (
                f"{augmented}\n\n"
                f"## Inherited from Previous Session\n"
                f"{self.inherited_context}"
            )

        system_msg = {"role": "system", "content": augmented}
        system_tokens = count_tokens(system_msg)
        budget = max_tokens - system_tokens

        # Walk backwards through messages, keeping as many as fit
        kept: list[dict] = []
        running = 0
        for msg in reversed(self.messages):
            tokens = count_tokens(msg)
            if running + tokens > budget:
                break
            kept.append(msg)
            running += tokens

        kept.reverse()
        return [system_msg, *kept]

    # ------------------------------------------------------------------
    # Serialisation
    # ------------------------------------------------------------------

    def to_dict(self) -> dict:
        """Serialise session for storage (JSON-safe)."""
        return {
            "session_id":        self.session_id,
            "user_id":           self.user_id,
            "created_at":        self.created_at,
            "last_activity":     self.last_activity,
            "state":             self.state.to_dict(),
            "messages":          list(self.messages),
            "inherited_context": self.inherited_context,
            "is_active":         self.is_active,
            # Summarizer layers
            "summarizer": {
                "verbatim":        [(u, a) for u, a in self.summarizer.verbatim],
                "layers":          list(self.summarizer.layers),
                "total_turns":     self.summarizer._total_turns,
                "verbatim_turns":  self.summarizer.verbatim_turns,
                "layer_size":      self.summarizer.layer_size,
                "layer_token_limit": self.summarizer.layer_token_limit,
            },
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Session":
        """Restore a session from a serialised dictionary."""
        session = cls(user_id=data["user_id"])
        session.session_id    = data["session_id"]
        session.created_at    = data["created_at"]
        session.last_activity = data["last_activity"]
        session.state_manager.state = ConversationState.from_dict(data["state"])
        session.messages      = data.get("messages", [])
        session.inherited_context = data.get("inherited_context")
        session.is_active     = data.get("is_active", True)

        # Restore summarizer
        s_data = data.get("summarizer", {})
        summ = ProgressiveSummarizer(
            verbatim_turns    = s_data.get("verbatim_turns", 5),
            layer_size        = s_data.get("layer_size", 10),
            layer_token_limit = s_data.get("layer_token_limit", 1500),
        )
        summ.verbatim      = [tuple(pair) for pair in s_data.get("verbatim", [])]
        summ.layers        = s_data.get("layers", [""] * ProgressiveSummarizer.NUM_LAYERS)
        summ._total_turns  = s_data.get("total_turns", 0)
        session.summarizer = summ

        return session


# ---------------------------------------------------------------------------
# SessionManager
# ---------------------------------------------------------------------------


class SessionManager:
    """Manage conversation sessions across users and time.

    Handles creation, expiry, branching, reset, and JSON persistence.

    Args:
        ttl_minutes:  Minutes of inactivity before a session expires.
        max_sessions: Maximum number of concurrent sessions kept in memory.
                      Oldest sessions are evicted when the limit is reached.

    Example::

        mgr = SessionManager(ttl_minutes=30)
        session = mgr.get_session("user-123")
        session.add_user_message("I need to book a flight")
    """

    def __init__(
        self,
        ttl_minutes: int = 60,
        max_sessions: int = 10_000,
    ) -> None:
        self.ttl_minutes  = ttl_minutes
        self.max_sessions = max_sessions
        self._sessions: dict[str, Session] = {}

    # ------------------------------------------------------------------
    # Core session access
    # ------------------------------------------------------------------

    def get_session(self, user_id: str) -> Session:
        """Return the active session for *user_id*, creating a new one if
        the session has expired or does not exist.

        Args:
            user_id: Application-level user identifier.

        Returns:
            An active :class:`Session`.
        """
        session = self._sessions.get(user_id)

        if session is not None and session.is_expired(self.ttl_minutes):
            session = None  # Treat as non-existent

        if session is None:
            session = self.create_session(user_id)

        return session

    def create_session(
        self,
        user_id: str,
        inherit_from: str | None = None,
    ) -> Session:
        """Create a (possibly new) session, optionally inheriting context.

        If *inherit_from* names a user whose session still exists, the new
        session carries forward the parent's user profile and goal summary.

        Args:
            user_id:      Owner of the new session.
            inherit_from: User ID whose session to inherit context from.

        Returns:
            The newly created :class:`Session`.
        """
        self._evict_if_needed()
        session = Session(user_id=user_id)

        if inherit_from and inherit_from in self._sessions:
            parent = self._sessions[inherit_from]
            session.state.user_name        = parent.state.user_name
            session.state.user_preferences = dict(parent.state.user_preferences)
            if parent.state.current_goal:
                session.inherited_context = (
                    f"Previous conversation goal: {parent.state.current_goal}. "
                    f"Completed steps: {', '.join(parent.state.subtasks_completed) or 'none'}."
                )

        self._sessions[user_id] = session
        return session

    def branch_session(
        self,
        user_id: str,
        context_keys: list[str] | None = None,
    ) -> Session:
        """Create a new session that inherits selected context from the current one.

        *context_keys* controls which context is carried forward:

        - ``"user_profile"``  — name, preferences
        - ``"goal_summary"``  — current goal and completed steps
        - ``"preferences"``   — user preferences only

        Args:
            user_id:      User whose session to branch from.
            context_keys: Which context types to inherit.  Defaults to
                          ``["user_profile", "goal_summary"]``.

        Returns:
            The new branched :class:`Session`.
        """
        context_keys = context_keys or ["user_profile", "goal_summary"]
        parent = self._sessions.get(user_id)

        self._evict_if_needed()
        new_session = Session(user_id=user_id)

        if parent:
            if "user_profile" in context_keys or "preferences" in context_keys:
                new_session.state.user_name = parent.state.user_name
                new_session.state.user_preferences = dict(parent.state.user_preferences)

            if "user_profile" in context_keys:
                new_session.state.user_provided_info = dict(
                    parent.state.user_provided_info
                )

            if "goal_summary" in context_keys and parent.state.current_goal:
                completed_str = (
                    ", ".join(parent.state.subtasks_completed)
                    if parent.state.subtasks_completed
                    else "none"
                )
                new_session.inherited_context = (
                    f"Previous conversation goal: {parent.state.current_goal}. "
                    f"Completed: {completed_str}."
                )

        self._sessions[user_id] = new_session
        return new_session

    def reset_session(
        self, user_id: str, keep_identity: bool = True
    ) -> Session:
        """Reset a user's session, optionally preserving user identity.

        Args:
            user_id:       User whose session to reset.
            keep_identity: When True, the new session retains *user_name*
                           and *user_preferences* from the old session.

        Returns:
            The fresh :class:`Session`.
        """
        old = self._sessions.get(user_id)
        self._evict_if_needed()
        session = Session(user_id=user_id)

        if keep_identity and old:
            session.state.user_name        = old.state.user_name
            session.state.user_preferences = dict(old.state.user_preferences)

        self._sessions[user_id] = session
        return session

    def end_session(self, user_id: str) -> None:
        """Explicitly end a session.

        The session is marked inactive and removed from memory.

        Args:
            user_id: User whose session to end.
        """
        session = self._sessions.pop(user_id, None)
        if session:
            session.is_active = False

    # ------------------------------------------------------------------
    # Housekeeping
    # ------------------------------------------------------------------

    def cleanup_expired(self) -> int:
        """Remove all expired sessions.

        Returns:
            The number of sessions removed.
        """
        expired = [
            uid for uid, s in self._sessions.items()
            if s.is_expired(self.ttl_minutes)
        ]
        for uid in expired:
            del self._sessions[uid]
        return len(expired)

    def get_active_sessions(self) -> int:
        """Return the count of sessions that have not yet expired."""
        return sum(
            1 for s in self._sessions.values()
            if not s.is_expired(self.ttl_minutes)
        )

    def _evict_if_needed(self) -> None:
        """Evict the oldest session when at capacity."""
        if len(self._sessions) >= self.max_sessions:
            oldest_uid = min(
                self._sessions,
                key=lambda uid: self._sessions[uid].last_activity,
            )
            del self._sessions[oldest_uid]

    # ------------------------------------------------------------------
    # Persistence
    # ------------------------------------------------------------------

    def persist_session(self, user_id: str) -> str:
        """Serialise a session to a JSON string.

        Args:
            user_id: User whose session to serialise.

        Returns:
            A JSON string, or ``"{}"`` if the user has no session.
        """
        session = self._sessions.get(user_id)
        if not session:
            return "{}"
        return json.dumps(session.to_dict())

    def restore_session(self, user_id: str, data: str) -> Session:
        """Deserialise a session from a JSON string.

        Args:
            user_id: User to assign the restored session to.
            data:    JSON string previously produced by :meth:`persist_session`.

        Returns:
            The restored :class:`Session`.
        """
        parsed = json.loads(data)
        session = Session.from_dict(parsed)
        # Ensure user_id is consistent
        session.user_id = user_id
        self._sessions[user_id] = session
        return session


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

def _demo() -> None:  # pragma: no cover
    """Demonstrate session management: creation, branching, reset, expiry."""

    print("=" * 70)
    print("SESSION MANAGER DEMO")
    print("=" * 70)

    mgr = SessionManager(ttl_minutes=30)

    # ── Create sessions for 3 users ─────────────────────────────────────
    print("\n── Creating sessions for 3 users ──")
    for user_id in ("alice", "bob", "carol"):
        s = mgr.get_session(user_id)
        print(f"  {user_id}: session {s.session_id[:8]}…")

    print(f"  Active sessions: {mgr.get_active_sessions()}")

    # ── Alice books a flight ─────────────────────────────────────────────
    print("\n── Alice's conversation ──")
    alice = mgr.get_session("alice")
    alice.state_manager.set_goal(
        "Book a flight to London",
        subtasks=["choose dates", "confirm payment"],
    )
    alice.add_user_message("I need to book a flight to London for June 15.")
    alice.add_agent_message("I can help! What's your return date?")
    alice.add_user_message("June 22. My name is Alice. My budget is $800.")
    alice.add_agent_message("Got it Alice — searching BA flights within $800.")
    alice.state_manager.mark_subtask_complete("choose dates")

    print(f"  Alice state:\n{alice.state.to_prompt_context()}")

    # ── Bob starts a different conversation ──────────────────────────────
    print("\n── Bob's conversation ──")
    bob = mgr.get_session("bob")
    bob.add_user_message("What's the weather like in Tokyo?")
    bob.add_agent_message("Tokyo is sunny and 22°C today.")
    print(f"  Bob messages: {len(bob.messages)}")

    # ── Session branching ────────────────────────────────────────────────
    print("\n── Alice branches to a new session (related task) ──")
    alice2 = mgr.branch_session("alice", context_keys=["user_profile", "goal_summary"])
    print(f"  New session ID: {alice2.session_id[:8]}…")
    print(f"  Inherited user name: {alice2.state.user_name}")
    print(f"  Inherited context: {alice2.inherited_context}")

    # ── Session reset ────────────────────────────────────────────────────
    print("\n── Bob resets his session (keep identity) ──")
    bob.state.user_name = "Bob"
    bob_reset = mgr.reset_session("bob", keep_identity=True)
    print(f"  New session ID: {bob_reset.session_id[:8]}…")
    print(f"  User name preserved: {bob_reset.state.user_name}")
    print(f"  Goal cleared: {bob_reset.state.current_goal!r}")
    print(f"  Messages cleared: {len(bob_reset.messages)}")

    # ── Session expiry ────────────────────────────────────────────────────
    print("\n── Session expiry simulation ──")
    carol = mgr.get_session("carol")
    carol_id_before = carol.session_id

    # Manually expire by backdating last_activity
    carol.last_activity = time.time() - (45 * 60)  # 45 minutes ago
    print(f"  Carol's session expired: {carol.is_expired(30)}")

    carol_new = mgr.get_session("carol")
    print(f"  Got new session after expiry: {carol_new.session_id[:8]}…")
    print(f"  Different from old: {carol_new.session_id != carol_id_before}")

    # ── Persist and restore ───────────────────────────────────────────────
    print("\n── Persistence roundtrip for Alice ──")
    json_str = mgr.persist_session("alice")
    print(f"  Serialised: {len(json_str)} bytes")

    mgr2 = SessionManager(ttl_minutes=30)
    restored = mgr2.restore_session("alice", json_str)
    print(f"  Restored session ID: {restored.session_id[:8]}…")
    print(f"  Restored goal: {restored.state.current_goal}")
    print(f"  Restored user name: {restored.state.user_name}")
    print(f"  Restored messages: {len(restored.messages)}")

    # Verify key fields match
    original = mgr.get_session("alice")
    assert restored.session_id == original.session_id
    assert restored.state.current_goal == original.state.current_goal
    assert restored.state.user_name == original.state.user_name
    print("  ✓ Restored state matches original")

    # ── Cleanup ───────────────────────────────────────────────────────────
    print("\n── Cleanup expired sessions ──")
    # Expire alice's sessions
    for uid in list(mgr._sessions):
        mgr._sessions[uid].last_activity = time.time() - (35 * 60)
    removed = mgr.cleanup_expired()
    print(f"  Sessions removed: {removed}")
    print(f"  Active sessions remaining: {mgr.get_active_sessions()}")

    print("\nDemo complete.")


if __name__ == "__main__":
    _demo()

File: code/nodejs/05-context-assembly/session_manager.ts
/**
 * Session Manager — multi-session conversation management.
 *
 * TypeScript port of code/python/05-context-assembly/session_manager.py
 *
 * Handles session creation, expiry, branching, reset, and JSON persistence.
 *
 * See: docs/04-context-engineering/04-multi-turn-context-management.md
 */

import { randomUUID } from "crypto";
import { countTokens } from "./context_budget.js";
import { ProgressiveSummarizer, SummarizerData } from "./progressive_summarizer.js";
import { ConversationState, StateManager } from "./state_manager.js";

// ---------------------------------------------------------------------------
// Session
// ---------------------------------------------------------------------------

export interface SessionData {
  sessionId:        string;
  userId:           string;
  createdAt:        number;
  lastActivity:     number;
  state:            ReturnType<ConversationState["toDict"]>;
  messages:         Array<{ role: string; content: string; tool_calls?: unknown[] }>;
  inheritedContext: string | null;
  isActive:         boolean;
  summarizer:       SummarizerData;
}

export class Session {
  readonly sessionId:  string;
  readonly userId:     string;
  readonly createdAt:  number;
  lastActivity:        number;
  readonly stateManager: StateManager;
  messages: Array<{ role: string; content: string; tool_calls?: unknown[] }> = [];
  readonly summarizer: ProgressiveSummarizer;
  inheritedContext: string | null = null;
  isActive: boolean = true;

  constructor(userId: string) {
    this.sessionId    = randomUUID();
    this.userId       = userId;
    this.createdAt    = Date.now() / 1000;
    this.lastActivity = Date.now() / 1000;
    this.stateManager = new StateManager();
    this.summarizer   = new ProgressiveSummarizer();
  }

  /** Convenience accessor for the session's ConversationState. */
  get state(): ConversationState {
    return this.stateManager.state;
  }

  // ------------------------------------------------------------------
  // Lifecycle
  // ------------------------------------------------------------------

  isExpired(ttlMinutes: number): boolean {
    return (Date.now() / 1000 - this.lastActivity) > ttlMinutes * 60;
  }

  touch(): void {
    (this as { lastActivity: number }).lastActivity = Date.now() / 1000;
  }

  // ------------------------------------------------------------------
  // Message management
  // ------------------------------------------------------------------

  addUserMessage(message: string): void {
    this.touch();
    this.messages.push({ role: "user", content: message });
    this.stateManager.processUserTurn(message);
  }

  async addAgentMessage(
    message: string,
    toolCalls?: Array<{ name?: string; function?: { name: string } }>,
  ): Promise<void> {
    this.touch();
    const msg: { role: string; content: string; tool_calls?: unknown[] } = {
      role: "assistant",
      content: message,
    };
    if (toolCalls) msg.tool_calls = toolCalls;
    this.messages.push(msg);
    this.stateManager.processAgentTurn(message, toolCalls);

    // Feed turn to summarizer — find the preceding user message
    let userContent = "";
    for (let i = this.messages.length - 2; i >= 0; i--) {
      if (this.messages[i].role === "user") {
        userContent = this.messages[i].content;
        break;
      }
    }
    await this.summarizer.addTurn(userContent, message);
  }

  buildMessagesForLlm(
    systemPrompt: string = "",
    maxTokens: number = 100_000,
  ): Array<{ role: string; content: string }> {
    let augmented = this.stateManager.buildSystemPromptWithState(systemPrompt);
    const summaryCtx = this.summarizer.getContext();
    if (summaryCtx)
      augmented += `\n\n## Conversation History Summary\n${summaryCtx}`;
    if (this.inheritedContext)
      augmented += `\n\n## Inherited from Previous Session\n${this.inheritedContext}`;

    const systemMsg = { role: "system", content: augmented };
    const systemTokens = countTokens(augmented);
    const budget = maxTokens - systemTokens;

    const kept: Array<{ role: string; content: string }> = [];
    let running = 0;
    for (let i = this.messages.length - 1; i >= 0; i--) {
      const msg = this.messages[i];
      const tokens = countTokens(msg.content ?? "");
      if (running + tokens > budget) break;
      kept.unshift({ role: msg.role, content: msg.content });
      running += tokens;
    }

    return [systemMsg, ...kept];
  }

  // ------------------------------------------------------------------
  // Serialisation
  // ------------------------------------------------------------------

  toDict(): SessionData {
    return {
      sessionId:        this.sessionId,
      userId:           this.userId,
      createdAt:        this.createdAt,
      lastActivity:     this.lastActivity,
      state:            this.state.toDict(),
      messages:         [...this.messages],
      inheritedContext: this.inheritedContext,
      isActive:         this.isActive,
      summarizer:       this.summarizer.toDict(),
    };
  }

  static fromDict(data: SessionData): Session {
    const session = new Session(data.userId);
    (session as { sessionId: string }).sessionId    = data.sessionId;
    (session as { createdAt: number }).createdAt    = data.createdAt;
    session.touch();
    (session as { lastActivity: number }).lastActivity = data.lastActivity;
    session.stateManager.state.currentGoal = ConversationState.fromDict(data.state).currentGoal;
    Object.assign(session.stateManager.state, ConversationState.fromDict(data.state));
    session.messages         = data.messages ?? [];
    session.inheritedContext = data.inheritedContext ?? null;
    (session as { isActive: boolean }).isActive = data.isActive ?? true;
    (session as { summarizer: ProgressiveSummarizer }).summarizer =
      ProgressiveSummarizer.fromDict(data.summarizer ?? {
        verbatim: [], layers: ["", "", ""], totalTurns: 0,
        verbatimTurns: 5, layerSize: 10, layerTokenLimit: 1500,
      });
    return session;
  }
}

// ---------------------------------------------------------------------------
// SessionManager
// ---------------------------------------------------------------------------

export class SessionManager {
  readonly ttlMinutes:  number;
  readonly maxSessions: number;
  private _sessions:    Map<string, Session> = new Map();

  constructor(ttlMinutes: number = 60, maxSessions: number = 10_000) {
    this.ttlMinutes  = ttlMinutes;
    this.maxSessions = maxSessions;
  }

  // ------------------------------------------------------------------
  // Core session access
  // ------------------------------------------------------------------

  getSession(userId: string): Session {
    const existing = this._sessions.get(userId);
    if (existing && !existing.isExpired(this.ttlMinutes))
      return existing;
    return this.createSession(userId);
  }

  createSession(userId: string, inheritFrom?: string): Session {
    this._evictIfNeeded();
    const session = new Session(userId);

    if (inheritFrom) {
      const parent = this._sessions.get(inheritFrom);
      if (parent) {
        session.state.userName        = parent.state.userName;
        session.state.userPreferences = { ...parent.state.userPreferences };
        if (parent.state.currentGoal) {
          session.inheritedContext =
            `Previous conversation goal: ${parent.state.currentGoal}. ` +
            `Completed steps: ${parent.state.subtasksCompleted.join(", ") || "none"}.`;
        }
      }
    }

    this._sessions.set(userId, session);
    return session;
  }

  branchSession(
    userId: string,
    contextKeys: string[] = ["user_profile", "goal_summary"],
  ): Session {
    const parent = this._sessions.get(userId);
    this._evictIfNeeded();
    const newSession = new Session(userId);

    if (parent) {
      if (contextKeys.includes("user_profile") || contextKeys.includes("preferences")) {
        newSession.state.userName        = parent.state.userName;
        newSession.state.userPreferences = { ...parent.state.userPreferences };
      }
      if (contextKeys.includes("user_profile")) {
        newSession.state.userProvidedInfo = { ...parent.state.userProvidedInfo };
      }
      if (contextKeys.includes("goal_summary") && parent.state.currentGoal) {
        const completed = parent.state.subtasksCompleted.join(", ") || "none";
        newSession.inheritedContext =
          `Previous conversation goal: ${parent.state.currentGoal}. ` +
          `Completed: ${completed}.`;
      }
    }

    this._sessions.set(userId, newSession);
    return newSession;
  }

  resetSession(userId: string, keepIdentity: boolean = true): Session {
    const old = this._sessions.get(userId);
    this._evictIfNeeded();
    const session = new Session(userId);

    if (keepIdentity && old) {
      session.state.userName        = old.state.userName;
      session.state.userPreferences = { ...old.state.userPreferences };
    }

    this._sessions.set(userId, session);
    return session;
  }

  endSession(userId: string): void {
    const session = this._sessions.get(userId);
    if (session) {
      (session as { isActive: boolean }).isActive = false;
      this._sessions.delete(userId);
    }
  }

  // ------------------------------------------------------------------
  // Housekeeping
  // ------------------------------------------------------------------

  cleanupExpired(): number {
    let count = 0;
    for (const [uid, session] of this._sessions) {
      if (session.isExpired(this.ttlMinutes)) {
        this._sessions.delete(uid);
        count++;
      }
    }
    return count;
  }

  getActiveSessions(): number {
    let count = 0;
    for (const session of this._sessions.values())
      if (!session.isExpired(this.ttlMinutes)) count++;
    return count;
  }

  private _evictIfNeeded(): void {
    if (this._sessions.size >= this.maxSessions) {
      let oldestUid = "";
      let oldestTime = Infinity;
      for (const [uid, s] of this._sessions) {
        if (s.lastActivity < oldestTime) {
          oldestTime = s.lastActivity;
          oldestUid  = uid;
        }
      }
      if (oldestUid) this._sessions.delete(oldestUid);
    }
  }

  // ------------------------------------------------------------------
  // Persistence
  // ------------------------------------------------------------------

  persistSession(userId: string): string {
    const session = this._sessions.get(userId);
    if (!session) return "{}";
    return JSON.stringify(session.toDict());
  }

  restoreSession(userId: string, data: string): Session {
    const parsed = JSON.parse(data) as SessionData;
    const session = Session.fromDict(parsed);
    session.userId !== userId && ((session as { userId: string }).userId = userId);
    this._sessions.set(userId, session);
    return session;
  }
}
